File: Flatten Binary Tree to Linked List/Flatten_Binary_Tree_to_Linked_List.cpp
// Runtime: 12 ms, faster than 100.00% of C++ online submissions for Flatten Binary Tree to Linked List.
// Memory Usage: 11 MB, less than 0.99% of C++ online submissions for Flatten Binary Tree to Linked List.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution 
{
public:
    void flatten(TreeNode* root) 
    {
        helper(root);
    }
private:
    TreeNode* helper(TreeNode *root)
    {
        if (root == 0 || (root->left == 0 && root->right == 0))
            return root;

        TreeNode *leftlastNode = helper(root->left);
        TreeNode *rightlastNode = helper(root->right);

        if (leftlastNode)
        {
            TreeNode *rightNode = root->right;
            root->right = root->left;
            root->left = 0;
            leftlastNode->right = rightNode;
        }

        if (rightlastNode == 0)
            return leftlastNode;
        else
            return rightlastNode;
    }
};

// 非递归版本的
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// Runtime: 12 ms, faster than 100.00% of C++ online submissions for Flatten Binary Tree to Linked List.
// Memory Usage: 11 MB, less than 0.99% of C++ online submissions for Flatten Binary Tree to Linked List.
class Solution 
{
public:
    void flatten(TreeNode* root) 
    {
        while (root)
        {
            if (root->left && root->right)
            {
                TreeNode *temp = root->left;
                while (temp->right)
                    temp = temp->right;
                temp->right = root->right;
            }

            if (root->left)
                root->right = root->left;
            
            root->left = 0;
            root = root->right;
        }
    }
};

File: Flatten Binary Tree to Linked List/Flatten_Binary_Tree_to_Linked_List.py
# Runtime: 44 ms, faster than 97.20% of Python3 online submissions for Flatten Binary Tree to Linked List.
# Memory Usage: 12.7 MB, less than 0.96% of Python3 online submissions for Flatten Binary Tree to Linked List.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: 'TreeNode') -> 'None':
        """
        Do not return anything, modify root in-place instead.
        """
        self.helper(root)
    def helper(self, root):
        if root is None or (root.left is None and root.right is None):
            return root
        leftlastNode = self.helper(root.left)
        rightlastNode = self.helper(root.right)

        if leftlastNode is not None:
            rightNode = root.right
            root.right = root.left
            root.left = None
            leftlastNode.right = rightNode
        
        return leftlastNode if rightlastNode is None else rightlastNode
